README
Devel::Isa::Explainer

This module is a simple tool for quickly visualizing inheritance
hierarchies to quickly see what functions are available for a given
class, or to ascertain where a given function you might see in use is
coming from.

This module does not concern itself with any of the fanciness of Roles,
and instead, relies entirely on standard Perl5 Object Model
infrastructure. ( Roles are effectively invisible at run-time as they
appear as composed functions in the corresponding class )

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Devel-Isa-Explainer

{{$NEXT}}
 - CLONE/CLONE_SKIP now exempt from shadowing logic. ( Closes #5 )

 [Dependencies::Stats]
 - Dependencies changed since 0.002900, see misc/*.deps* for details
 - develop: +2
 - runtime: (recommends: +1)

0.002900 2016-06-26T19:35:43Z 7405ed0
 - UNIVERSAL now automatically shown in inheritance. ( Closes #11 )
 - discovering subs using "can" now removed, as "can" was prematurely exposing subs in crossed inheritances, which
   leads to over-zealous de-duplication, and additional appearances of shadowing where no actual shadowing was
   happening. ( #10 )
 - blessed coderefs no longer misunderstood.
 - deduplicating subs now entirely removed, subs that are actually there in the package must always be displayed as
   they are either injected by roles or importing, and can actually end up calling themselves under inheritance. ( This
   however is highly prone to explosions because MRO core sucks ) ( Related to #10 )
 - isa-splain now supports --help and --version ( Closes #9 )

 [Dependencies::Stats]
 - Dependencies changed since 0.002001, see misc/*.deps* for details
 - runtime: +2 -1
 - test: +1

 [Documentation]
 - Improve bin/isa-splain's documentation to be more helpful in places where `man isa-splain` might be expected to
   work.

 [Internals]
 - Highligher functions now receive all sub metadata directly as a hash, instead of being called with a list of binary
   flags.
 - Parents of classes now recorded in internal data structures, but not yet exposed.
 - MRO type of classes now recorded in internal data structures.
 - XSUB, Constant and stub-subs marked in internal data structures.

0.002001 2016-05-19T14:07:12Z c167598
 [Documentation]
 - Reference updated screenshots.

 [Internals]
 - use namespace::clean in a few more places.

0.002000 2016-05-19T12:30:00Z 6c368ff
 - isa-splain now accepts -M parameter to explicitly specify modules to load. This is helpful in handling cases where
   one module inlines others, ( B -> B::CV ) or a module can't be loaded on its own ( Class::MOP::Class )

 [Bug #6]
 - Poking MAX_WIDTH to a value smaller than 5 no longer triggers an infinite loop, instead, a minimum of one package
   per line will be emitted, even if it exceeds MAX_WIDTH.

 [Dependencies::Stats]
 - Dependencies changed since 0.001001, see misc/*.deps* for details
 - develop: +1
 - runtime: +1

 [Internals]
 - factored out _class_subs function
 - Method Resolution mechanics reworked to more closely reflect how perl-itself views methods. This will reduce certain
   redundant duplicate coderefs and only show them at the lowest point in the heirachy where they are defined.
 - Now fatals when inheritance lookup returns nothing and the package had a compile failure.

0.001001 2016-04-01T09:31:35Z ab66fba
 - Minor documentation fixes.

0.001000 2016-03-31T13:55:56Z 3c0283f
 - First version.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.047.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Pretty Print Hierarchies of Subs in Packages",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Devel-Isa-Explainer",
  "EXE_FILES" => [
    "bin/isa-splain"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Devel::Isa::Explainer",
  "PREREQ_PM" => {
    "B" => 0,
    "Carp" => 0,
    "Exporter" => 0,
    "MRO::Compat" => 0,
    "Module::Load" => 0,
    "Scalar::Util" => 0,
    "Term::ANSIColor" => "3.00",
    "constant" => "1.03",
    "namespace::clean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::Builder" => 0,
    "Test::Differences" => 0,
    "Test::More" => "0.89"
  },
  "VERSION" => "0.002901",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t t/cli/*.t t/internals/*.t t/internals/mro/*.t"
  }
);


my %FallbackPrereqs = (
  "B" => 0,
  "Carp" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "MRO::Compat" => 0,
  "Module::Load" => 0,
  "Scalar::Util" => 0,
  "Term::ANSIColor" => "3.00",
  "Test::Builder" => 0,
  "Test::Differences" => 0,
  "Test::More" => "0.89",
  "constant" => "1.03",
  "namespace::clean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/App/Isa/Splain.pm
use 5.006;    # our
use strict;
use warnings;

package App::Isa::Splain;

our $VERSION = '0.002901'; # TRIAL

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

# ABSTRACT: Visualize Module Hierarchies on the command line

use Module::Load qw( load );
use Carp qw( croak );
use Devel::Isa::Explainer qw( explain_isa );

# Perl critic is broken. This is not a void context.
## no critic (BuiltinFunctions::ProhibitVoidMap)
use constant 1.03 ( { map { ( ( sprintf '_E%x', $_ ), ( sprintf ' E<%s#%d>', __PACKAGE__, $_ ) ) } 1 .. 3 } );

use namespace::clean;











sub new {
  my ( $class, @args ) = @_;
  my (%args) = ref $args[0] ? %{ $args[0] } : @args;
  return bless \%args, $class;
}












sub new_from_ARGV {
  my (@args) = defined $_[1] ? @{ $_[1] } : @ARGV;
  my $module;
  my @load_modules;
  while ( @args ) {
    my $argument = shift @args;
    if ( not defined $module and $argument !~ /\A-/sx ) {
      $module = $argument;
      next;
    }
    if( $argument =~ /\A-M(.*)\z/sx ) {
      push @load_modules, $1;
      next;
    }
    ## no critic (RequireCheckedSyscalls)
    if( '--help' eq $argument) {
      print _help();
      exit;
    }
    if( '--version' eq $argument ) {
      print _version();
      exit;
    }
    croak 'Unexpected argument ' . $argument . _E3 . qq[\nSee $0 --help for more information];
  }
  defined $module or croak 'Expected a module name, got none' . _E1 . qq[\nSee $0 --help for more information];
  return $_[0]->new( module => $module, load_modules => [ @load_modules ? @load_modules : $module ] );
}

sub _load_modules { return @{ $_[0]->{load_modules} } }
sub _module { return $_[0]->{module} }
sub _output { return ( $_[0]->{output} || *STDOUT ) }







sub run {
  my ($self) = @_;
  load $_ for $self->_load_modules;
  croak "Could not print to output handle: $! $^E" . _E2
    unless print { $self->_output } explain_isa( $self->_module );

  return 0;
}

sub _help {
  return <<"EOF";
Usage: $0 [OPTIONS] MODNAME
Load and inspect MODNAME's ISA Inheritance.

OPTIONS:
  -MLoad::Module    - Load "Load::Module" instead of MODNAME ( May be
                      specified multiple times )

  --help            - Show this help and exit
  --version         - Show version and exit
EOF
}

sub _version {
  my $pkg = __PACKAGE__;
  return <<"EOF";
$0 ($pkg) $VERSION
EOF
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

App::Isa::Splain - Visualize Module Hierarchies on the command line

=head1 VERSION

version 0.002901

=head1 SYNOPSIS

  my $instance = App::Isa::Splain->new_from_ARGV;
  $instance->run;

=for html <center><img alt="Colorised output from a Moose::Meta::Class" src="http://kentnl.github.io/screenshots/Devel-Isa-Explainer/1/c1.png" width="820" height="559" /></center>

=head1 METHODS

=head2 C<new>

Creates an Explainer script for the given module

  my $instance = App::Isa::Splain->new(
    module => "module::name"
  );

=head2 C<new_from_ARGV>

Creates an Explainer script by passing command line arguments

  my $instance = App::Isa::Splain->new_from_ARGV;
  my $instance = App::Isa::Splain->new_from_ARGV(\@ARGV); # Alternative syntax

See L<COMMAND LINE ARGUMENTS|/COMMAND LINE ARGUMENTS>

=head2 C<run>

Executes the explainer and prints its output.

=head1 COMMAND LINE ARGUMENTS

  isa-splain [-MModule::Name] Module::Name

=over 4

=item * C<Module::Name>

A module to C<require> and analyze the C<ISA> of.

=item * C<-MI<Module::Name>>

A module to load instead of the module being analyzed, for example:

  isa-splain -MB B::CV
  isa-splain -Moose Class::MOP::Class

Helpful for cases where simple C<isa-splain Module::Name> causes problems.

=back

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (qw( explainer Explainer interoperability DIAGNOSTICS ));
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');
$bundle->remove_policy('TestingAndDebugging::ProhibitNoStrict');
$bundle->remove_policy('Subroutines::ProhibitCallsToUndeclaredSubs');
$bundle->remove_policy('Subroutines::RequireArgUnpacking');
#$bundle->remove_policy('CodeLayout::RequireUseUTF8');
#$bundle->remove_policy('ErrorHandling::RequireCarping');
$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



lib/Devel/Isa/Explainer/_MRO.pm
use 5.006;    # our
use strict;
use warnings;

package Devel::Isa::Explainer::_MRO;

# ABSTRACT: Method-resolution-order Utilities for DIE

# AUTHORITY

our $VERSION = '0.002901';

use MRO::Compat ();
use Exporter    ();
use Scalar::Util qw(reftype);

BEGIN {
  ## no critic (ProhibitCallsToUnexportedSubs)
  *import              = \&Exporter::import;
  *_mro_get_linear_isa = \&mro::get_linear_isa;
  *_mro_is_universal   = \&mro::is_universal;
}

our @EXPORT_OK = qw(
  is_mro_proxy
  get_linear_isa
  get_package_sub
  get_package_subs
  get_linear_class_shadows
  get_parents
  get_linear_method_map
  get_linear_class_map
  get_flattened_class
);

our %SHADOW_EXEMPT = (
  map { $_ => 1 } (

    # http://perldoc.perl.org/perlmod.html#Making-your-module-threadsafe
    # CLONE is called at all levels, shadowed or not
    'CLONE',

    # CLONE_SKIP is also called on all levels, shadowed or not.
    ( $] >= 5.008007 ? 'CLONE_SKIP' : () ),
  )
);

BEGIN {
  # MRO Proxies removed since 5.009_005
  *MRO_PROXIES = ( $] <= 5.009005 ) ? sub() { 1 } : sub() { 0 };
}

use namespace::clean -except => 'import';

=func is_mro_proxy

  if ( MRO_PROXIES and is_mro_proxy( $package, $sub ) ) {
    // its a proxy
  } else {
    // anything else
  }

Prior to 5.009_005, L<< backwards-compatibility support for C<MRO>|MRO::Compat >> for
5.8 has to install "proxy" subs at various levels that I<emulate> alternative
resolution orders by hiding relevant nodes in the gaps in tree.

This detects those nodes so that we can pretend they don't exist.

=cut

sub is_mro_proxy {

  # Note: this sub should be optimised out from calling anyway
  # but this is just a failsafe
  MRO_PROXIES ? !!( $Class::C3::MRO{ $_[0] } || {} )->{methods}{ $_[1] } : 0;
}

=func get_linear_isa

  my $isa = get_linear_isa( $class );

This function is like C<< mro::get_linear_isa()|mro/get_linear_isa >>, with
the exception that it includes C<UNIVERSAL> and any parents of C<UNIVERSAL>
where relevant.

If pointed at C<UNIVERSAL>, will include C<UNIVERSAL>s parents.

If pointed at a L<< parent of C<UNIVERSAL>|mro/is_univeral >>, will B<not>
show C<UNIVERSAL>, despite the fact calling C<< ->can() >> on a parent of
C<UNIVERSAL> still works, despite the fact its actually defined in C<UNIVERSAL>.

=cut

sub get_linear_isa {
  [
    @{ _mro_get_linear_isa( $_[0] ) },
    #<<<
    _mro_is_universal( $_[0] )
      ? ()
      : @{ _mro_get_linear_isa('UNIVERSAL') },
    #>>>
  ];
}

=func get_package_sub

  my $sub = get_package_sub($package, $sub);

Fetch a directly defined C<CodeRef> from C<$package> named C<$sub>

Fake proxy methods (such as Class::C3 proxies) and stubs are ignored by this
and instead return C<undef>

  $result = undef / CODEREF

=cut

sub get_package_sub {
  return undef if MRO_PROXIES and is_mro_proxy(@_);
  my ( $package, $sub ) = @_;

  # this is counter intuitive, but literally
  # everything in a stash that is not a glob *is* a sub.
  #
  # Though they're usually constant-subs.
  #
  # Globs however can /contain/ subs in their {CODE} slot,
  # but globs are not subs.
  my $namespace = do {
    no strict 'refs';
    \%{"${package}::"};
  };
  return undef unless exists $namespace->{$sub};
  if ( 'GLOB' eq reftype \$namespace->{$sub} ) {

    # Autoviv guard.
    return defined *{ \$namespace->{$sub} }{'CODE'} ? *{ \$namespace->{$sub} }{'CODE'} : undef;
  }

  # Note: This vivifies the stash slot into a glob...
  # there's not much that can be done about this at present.
  # Package::Stash does the same.
  #
  # This means the first of us or Package::Stash to traverse a symtable turns
  # everything into globs in order to get coderefs out.
  #
  # Ideally, we don't do this, but ENEEDINFO
  return \&{"${package}::${sub}"};
}

=func get_package_subs

  my $hashref = get_package_subs( $packagename );

Returns a hash of the packages directly defined C<sub>'s.

  $result = { SUBNAME => CODEREF, ... };

=cut

# like get_package_sub, but does a whole class at once and returns a hashref
# of { name => CODEREF }
sub get_package_subs {
  my ($package)   = @_;
  my ($namespace) = do {
    no strict 'refs';
    \%{"${package}::"};
  };
  my (@symnames) = do {
    no strict 'refs';
    keys %{"${package}::"};
  };
  my $subs = {};
  for my $symname (@symnames) {

    my $reftype = reftype \$namespace->{$symname};

    # Globs are only subs if they contain a CODE slot
    # all non-globs vivify to subs.
    # Order can't be changed though, because the second test requires the
    # first to be true to test, so defined is only tested when eq.
    next if ( 'GLOB' eq $reftype ) and not defined *{ \$namespace->{$symname} }{'CODE'};
    next if MRO_PROXIES and is_mro_proxy( $package, $symname );
    $subs->{$symname} =
      'GLOB' eq $reftype
      ? *{ \$namespace->{$symname} }{'CODE'}
      : \&{"${package}::${symname}"};
  }
  $subs;
}

=func get_linear_class_shadows

  my $arrayref = get_linear_class_shadows( $classname )

Combines C<get_linear_isa()> and C<get_package_subs()>,
traversing the inheritance bottom up, computing shadowing
as it goes.

Returns:

  $result     = [ $hashref, $hashref, $hashref,   ... ]
  $hashrefref = { class => CLASSNAME, subs => $submap }
  $submap     = { SUBNAME => $subrecord,          ... }
  $subrecord  = { shadowing => BOOLEAN,
                  shadowed  => BOOLEAN,
                  ref       => CODEREF,               }

=cut

sub get_linear_class_shadows {
  my ($class) = @_;

  # Contains the "image" made bottom up
  # for comparison/detecting shadows.
  my $methods = {};
  my @isa_out;
  for my $package ( reverse @{ get_linear_isa($class) } ) {
    my $subs = get_package_subs($package);
    my $node = {};
    for my $subname ( keys %{$subs} ) {

      # first node is never shadowing
      if ( not exists $methods->{$subname} ) {
        $node->{$subname} = { shadowing => 0, shadowed => 0, ref => $subs->{$subname} };

        # Contains a reference to the previous incarnation
        # for later modification
        $methods->{$subname} = $node->{$subname};
        next;
      }
      if ( exists $SHADOW_EXEMPT{$subname} ) {
        $node->{$subname} = { shadowing => 0, shadowed => 0, ref => $subs->{$subname} };
      }
      else {
        $node->{$subname} = { shadowing => 1, shadowed => 0, ref => $subs->{$subname} };
        $methods->{$subname}->{shadowed} = 1;    # mark previous version shadowed
      }
      $methods->{$subname} = $node->{$subname};    # update current
    }
    unshift @isa_out, { class => $package, subs => $node };
  }
  \@isa_out;
}

=func get_parents

  my $parents = get_parents( $package );

This utility finds the effective "depth 1" parents of a given class.
That is, in normal conditions, it just returns the contents of C<@ISA> verbatim.

However, if C<@ISA> is empty, it returns the effective parent, C<UNIVERSAL>
unless of course, the given class is a parent of C<UNIVERSAL> itself (insert drugs here)
at which point it will return an empty list.

Because despite the fact a parent of C<UNIVERSAL> can call C<UNIVERSAL> methods,
reporting C<< UNIVERSAL->parent->parent == UNIVERSAL >> will of course create cycles
for anyone who touches it.

=cut

sub get_parents {
  my ($package) = @_;
  my $namespace = do {
    no strict 'refs';
    \%{"${package}::"};
  };

  if ( exists $namespace->{ISA} ) {
    my $entry_ref = \$namespace->{ISA};
    if (  'GLOB' eq reftype $entry_ref
      and defined *{$entry_ref}{ARRAY}
      and @{ *{$entry_ref}{ARRAY} } )
    {
      return [ @{ *{$entry_ref}{ARRAY} } ];
    }
  }
  return [] if _mro_is_universal($package);
  ['UNIVERSAL'];
}

=func get_linear_method_map

  my $arrayref = get_linear_method_map( $classname, $method )

Returns an C<ArrayRef> describing the vertical stack of a given method.

C<ISA> levels without defined C<CodeRefs> are represented as C<undef>

  $result   = [ $arrayref, $arrayref, $arrayref, ... ]
  $arrayref = [ CLASSNAME, undef / CODEREF           ]

=cut

sub get_linear_method_map {
  my ( $class, $method ) = @_;
  return [ map { [ $_, get_package_sub( $_, $method ) ] } @{ get_linear_isa($class) } ];
}

=func get_linear_class_map

  my $arrayref = get_linear_class_map( $classname )

Returns C<CodeRef> stashes for all packages in C<$classname>'s inheritance (including C<UNIVERSAL>s)
in method-resolution-order.

Returns:

  $result   = [ $arrayref, $arrayref, $arrayref,  ... ]
  $arrayref = [ CLASSNAME, $submap                    ]
  $submap   = { SUBNAME => CODEREF,               ... }

=cut

sub get_linear_class_map {
  my ($class) = @_;
  [ map { [ $_, get_package_subs($_) ] } @{ get_linear_isa($class) } ];
}

=func get_flattened_class

  my $hashref = get_flattened_class( $class_name );

Returns a fully expanded "Flat" representation of a classes hierarchy,
with still enough data present to trace method resolution.

Returns:

  $result = { SUBNAME => $entry, ... }
  $entry  = { ref     => CODEREF,
              via     => CLASSNAME,
              parents => $parentrefs, }

  $parentrefs      = [ $parentref_entry, ... ]
  $parentref_entry = [ CLASSNAME, CODEREF    ]

=cut

sub get_flattened_class {
  my ($class) = @_;
  my $methods = {};
  for my $package ( reverse @{ get_linear_isa($class) } ) {
    my $subs = get_package_subs($package);
    for my $subname ( keys %{$subs} ) {
      $methods->{$subname}->{parents} ||= [];
      unshift @{ $methods->{$subname}->{parents} }, [ $methods->{$subname}->{via}, $methods->{$subname}->{ref} ]
        if exists $methods->{$subname}->{ref};
      $methods->{$subname}->{ref} = $subs->{$subname};
      $methods->{$subname}->{via} = $package;
    }
  }
  $methods;
}

1;
